### ISSUE TITLE
Memory Requirements for Deep Neural Networks

### ISSUE THREAD

**Issue by YunchaoYang** [YunchaoYang/Blogs#13]
# Estimating neural network memory consumption

### convolutional layer 
- n-th layer
  - weight = depth_n x (kernel_width x kernel_height) * depth_(n-1)
  - bias = depth_n

### fully connected layer
- n-th layer
  - weight = #outputs x #inputs
  - bias = #outputs

##  Layer blob size 
This is the output of each layer, which is saved to backprop gradient calculation in autograph.

###  n-th Conv layer
=width x height x depth
### FC layers 
=#outputs

## Total memory requirements  (train time)
sum of 
- Memory for parameters
- Memory for param gradients
- Memory for momentum

and 
- Memory for layer outputs
- Memory for layer error

plus 
- implementation overhead

#  Estimating neural network computation (FLOP/s)
Estimating computational complexity
## Convolution layer
Product of:
- Spatial width of the map
- Spatial height of the map
- Previous layer depth
- Current layer depth
- Kernel width
- Kernel height
## FC layer
= #inputs x #outputs

# Examples:
## resnet50
- 48 conv-layer, 1 maxpool, 1 avg pool
- number of parameters: 25.6 M
- size (MB) = 98 (keras ~25.6*4)
- FLOPs = 3.8 x10^9

# reference: 
1. http://imatge-upc.github.io/telecombcn-2016-dlcv/slides/D2L1-memory.pdf
